### ISSUE TITLE
Unit-like structs should not trigger `exhaustive_structs`

### ISSUE THREAD

**Issue by kpreid** [rust-lang/rust-clippy#12675]
### Summary

By using [unit-like struct syntax](https://doc.rust-lang.org/reference/items/structs.html), one is opting into behavior which is incompatible with ever adding any fields (the constant named after the struct). Therefore, adding `#[non_exhaustive]` does not provide any forward-compatibility benefit — adding fields is always a breaking change. Therefore, Clippy should not recommend `#[non_exhaustive]` on unit-like structs.

### Lint Name

exhaustive_structs

### Reproducer

I tried this code:

```rust
#![warn(clippy::exhaustive_structs)]
pub struct Foo;
```

I saw this happen:

```
warning: exported structs should not be exhaustive
 --> src/lib.rs:2:1
  |
2 | pub struct Foo;
  | ^^^^^^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#exhaustive_structs
note: the lint level is defined here
 --> src/lib.rs:1:9
  |
1 | #![warn(clippy::exhaustive_structs)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^
help: try adding #[non_exhaustive]
  |
2 + #[non_exhaustive]
3 | pub struct Foo;
  |
```

I expected to see this happen: No lint

### Version

```text
rustc 1.77.1 (7cf61ebde 2024-03-27)
binary: rustc
commit-hash: 7cf61ebde7b22796c69757901dd346d0fe70bd97
commit-date: 2024-03-27
host: x86_64-apple-darwin
release: 1.77.1
LLVM version: 17.0.6
```

**Comment by Alexendoo** [comment]
Is there a case `#[non_exhaustive]` doesn't catch? My understanding is that it prevents using the struct in a way that would reveal it's a unit struct

**Comment by kpreid** [comment]
>  My understanding is that it prevents using the struct in a way that would reveal it's a unit struct

Ah, I was not aware of this:

```rust
#[non_exhaustive]
pub struct Foo;
```

```rust
use scratchpad::Foo;
fn main() {
    let _: Foo = Foo;
}
```

```rust
error[E0423]: expected value, found struct `Foo`
 --> src/bin/scratchbin.rs:3:18
  |
3 |     let _: Foo = Foo;
  |                  ^^^ constructor is not visible here due to private fields
```

Sorry for filing an issue with an incorrect assumption.
